### ISSUE TITLE
test_add_new_node_while_add_new_table_during_bootstrapping fails to insert data during bootstrap (with raft disabled)

### ISSUE THREAD

**Issue by fruch** [scylladb/scylladb#14881]
test is trying to insert data during a node bootstrap with CL=ONE, and failing like the following:
```
self = <update_cluster_layout_tests.TestUpdateClusterLayout object at 0x7fb920806f90>

    def test_add_new_node_while_add_new_table_during_bootstrapping(self):
>       self._add_new_node_while_add_new_table("during")

update_cluster_layout_tests.py:1508: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
update_cluster_layout_tests.py:1499: in _add_new_node_while_add_new_table
    t.result()
/usr/lib64/python3.11/concurrent/futures/_base.py:449: in result
    return self.__get_result()
/usr/lib64/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
/usr/lib64/python3.11/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
update_cluster_layout_tests.py:1476: in run
    session.execute(insert)
cassandra/cluster.py:2700: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="Operation failed for ks1.cf1 - received 1 responses and 1 failures from 2 CL=ONE." info={'consistency': 'ONE', 'required_responses': 2, 'received_responses': 1, 'failures': 1}

cassandra/cluster.py:5019: WriteFailure
```

according to our elasticsearch data, it happened only once so far

### Logs:
https://jenkins.scylladb.com/job/scylla-master/job/next/6317/artifact/dtest/logs-full.release.001/1690560531266_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/

**Comment by DoronArazii** [comment]
@eliransin / @ptrsmrn please assign someone to triage this

**Comment by mykaul** [comment]
Where are the logs?

**Comment by ptrsmrn** [comment]
@fruch @DoronArazii 
I was able to run this dtest locally and it passed for me for both the OSS and enterprise - but I run it only once/twice. I think I will run it in a loop for OSS until it fails and start investigating from there.

```
(.venv) [piotrs@fedora scylla-dtest]$ SCYLLA_PRODUCT=scylla-enterprise pytest --scylla-version=unstable/enterprise:latest  update_cluster_layout_tests.py::TestUpdateClusterLayout::test_add_new_node_while_add_new_table_during_bootstrapping
S3 download: scylla-enterprise-unified-2023.3.0~dev-0.20230813.99a0316b5982.x86_64.tar.gz100%|██████████| 173M/173M [00:11<00:00, 14.5MB/s] 
Extracting /tmp/ccm-6lzhqntg.tar.gz (http://s3.amazonaws.com/downloads.scylladb.com/unstable/scylla-enterprise/enterprise/relocatable/2023-08-14T02:38:55Z/scylla-enterprise-unified-2023.3.0~dev-0.20230813.99a0316b5982.x86_64.tar.gz, /tmp/tmpbceplaq4) as version unstable/enterprise/latest ...
Relocatable package format version 3.0 detected.
Completed to install Scylla in the folder '/home/piotrs/.ccm/scylla-repository/unstable/enterprise/latest'
================================================================== test session starts ==================================================================
platform linux -- Python 3.11.3, pytest-7.4.0, pluggy-1.2.0
Using --randomly-seed=3738989344
rootdir: /home/piotrs/src/local_dtest/scylla-dtest
configfile: pytest.ini
plugins: xdist-3.3.1, timeout-2.1.0, subtests-0.11.0, repeat-0.9.1, randomly-3.13.0, metadata-3.0.0, elk-reporter-0.2.1, html-3.2.0
timeout: 2700.0s
timeout method: signal
timeout func_only: False
collected 1 item                                                                                                                                        

update_cluster_layout_tests.py::TestUpdateClusterLayout::test_add_new_node_while_add_new_table_during_bootstrapping PASSED                        [100%]

============================================================= 1 passed in 161.82s (0:02:41) =============================================================
(.venv) [piotrs@fedora scylla-dtest]$ SCYLLA_PRODUCT=scylla pytest --scylla-version=unstable/master:latest  update_cluster_layout_tests.py::TestUpdateClusterLayout::test_add_new_node_while_add_new_table_during_bootstrapping
S3 download: scylla-unified-5.4.0~dev-0.20230812.d1d1b6cf6e01.x86_64.tar.gz100%|██████████| 176M/176M [00:10<00:00, 17.3MB/s] 
Extracting /tmp/ccm-gnl3of50.tar.gz (http://s3.amazonaws.com/downloads.scylladb.com/unstable/scylla/master/relocatable/2023-08-12T03:07:56Z/scylla-unified-5.4.0~dev-0.20230812.d1d1b6cf6e01.x86_64.tar.gz, /tmp/tmpud_7v4a5) as version unstable/master/latest ...
Relocatable package format version 3.0 detected.
Completed to install Scylla in the folder '/home/piotrs/.ccm/scylla-repository/unstable/master/latest'
================================================================== test session starts ==================================================================
platform linux -- Python 3.11.3, pytest-7.4.0, pluggy-1.2.0
Using --randomly-seed=2404181767
rootdir: /home/piotrs/src/local_dtest/scylla-dtest
configfile: pytest.ini
plugins: xdist-3.3.1, timeout-2.1.0, subtests-0.11.0, repeat-0.9.1, randomly-3.13.0, metadata-3.0.0, elk-reporter-0.2.1, html-3.2.0
timeout: 2700.0s
timeout method: signal
timeout func_only: False
collected 1 item                                                                                                                                        

update_cluster_layout_tests.py::TestUpdateClusterLayout::test_add_new_node_while_add_new_table_during_bootstrapping PASSED                        [100%]

============================================================= 1 passed in 162.17s (0:02:42) =============================================================
(.venv) [piotrs@fedora scylla-dtest]$

**Comment by DoronArazii** [comment]
Thanks for the update Piotr.

**Comment by ptrsmrn** [comment]
I run it 300+ locally and it passed every time. 
We can either close this Issue as "not-reproducible" (also because the logs link expired) or I can try reproducing it by other means: maybe different dtests settings for example.

**Comment by fruch** [comment]
I'll take a look tomorrow on the history of this test, if it's a one off, we'll close it.

**Comment by fruch** [comment]
> I'll take a look tomorrow on the history of this test, if it's a one off, we'll close it.

seems like it failure with error only once, since the test introduce.
Closing it until we'll see it happening again

**Comment by avikivity** [comment]
Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6651/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_add_new_node_while_add_new_table_during_bootstrapping/

```
self = <update_cluster_layout_tests.TestUpdateClusterLayout object at 0x7f071413c4d0>

    @pytest.mark.no_boot_speedups
    def test_add_new_node_while_add_new_table_during_bootstrapping(self):
>       self._add_new_node_while_add_new_table("during")

update_cluster_layout_tests.py:1523: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
update_cluster_layout_tests.py:1513: in _add_new_node_while_add_new_table
    t.result()
/usr/lib64/python3.11/concurrent/futures/_base.py:449: in result
    return self.__get_result()
/usr/lib64/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
/usr/lib64/python3.11/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
update_cluster_layout_tests.py:1490: in run
    session.execute(insert)
cassandra/cluster.py:2708: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="Operation failed for ks1.cf1 - received 1 responses and 1 failures from 2 CL=ONE." info={'consistency': 'ONE', 'required_responses': 2, 'received_responses': 1, 'failures': 1}

cassandra/cluster.py:5048: WriteFailure
```

**Comment by avikivity** [comment]
Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6663/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_add_new_node_while_add_new_table_during_bootstrapping/

**Comment by avikivity** [comment]
```
cassandra.WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="Operation failed for ks1.cf1 - received 1 responses and 1 failures from 2 CL=ONE." info={'consistency': 'ONE', 'required_responses': 2, 'received_responses': 1, 'failures': 1}
```

```
self = <update_cluster_layout_tests.TestUpdateClusterLayout object at 0x7fa56f9574d0>

    @pytest.mark.no_boot_speedups
    def test_add_new_node_while_add_new_table_during_bootstrapping(self):
>       self._add_new_node_while_add_new_table("during")

update_cluster_layout_tests.py:1523: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
update_cluster_layout_tests.py:1513: in _add_new_node_while_add_new_table
    t.result()
/usr/lib64/python3.11/concurrent/futures/_base.py:449: in result
    return self.__get_result()
/usr/lib64/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
/usr/lib64/python3.11/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
update_cluster_layout_tests.py:1490: in run
    session.execute(insert)
cassandra/cluster.py:2708: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.WriteFailure: Error from server: code=1500 [Replica(s) failed to execute write] message="Operation failed for ks1.cf1 - received 1 responses and 1 failures from 2 CL=ONE." info={'consistency': 'ONE', 'required_responses': 2, 'received_responses': 1, 'failures': 1}

cassandra/cluster.py:5048: WriteFailure
```

**Comment by avikivity** [comment]
If CL=ONE, why do we wait for two responses? I guess RF=1 and we're bootstrapping a new node.

**Comment by mykaul** [comment]
Test log (https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/dtest-gw2.log):
```
9:02:32,594 787     update_cluster_layout_tests    DEBUG    update_cluster_layout_tests.py:377  | test_add_new_node_while_add_new_table_during_bootstrapping: Bootstrap node4: started streaming/repair
09:02:32,609 787     cassandra.protocol             WARNING  libevreactor.py     :370  | test_add_new_node_while_add_new_table_during_bootstrapping: Server warning: Using Replication Factor replication_factor=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.
09:02:55,779 787     cassandra.policies             INFO     policies.py         :289  | test_add_new_node_while_add_new_table_during_bootstrapping: Using datacenter 'datacenter1' for DCAwareRoundRobinPolicy (via host '127.0.39.4:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
09:02:55,779 787     cassandra.cluster              INFO     dtest_setup.py      :437  | test_add_new_node_while_add_new_table_during_bootstrapping: New Cassandra host <Host: 127.0.39.3:9042 datacenter1> discovered
09:02:55,779 787     cassandra.cluster              INFO     dtest_setup.py      :437  | test_add_new_node_while_add_new_table_during_bootstrapping: New Cassandra host <Host: 127.0.39.1:9042 datacenter1> discovered
09:02:55,780 787     cassandra.cluster              INFO     dtest_setup.py      :437  | test_add_new_node_while_add_new_table_during_bootstrapping: New Cassandra host <Host: 127.0.39.2:9042 datacenter1> discovered
09:02:56,206 787     errors                         ERROR    conftest.py         :225  | test_add_new_node_while_add_new_table_during_bootstrapping: test failed: 
```

node1 (https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/1697360579238_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/node1.log):
```
INFO  2023-10-15 09:02:43,732 [shard 0:stre] storage_service - bootstrap[dfa4d19e-229a-4670-b154-d3e0196379fc]: Marked ops done from coordinator=127.0.39.4
INFO  2023-10-15 09:02:43,959 [shard 0:goss] storage_service - Set host_id=74df4b09-7f6e-43e4-9bbe-88ce5962f30d to be owned by node=127.0.39.4
```

node2 (https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/1697360579238_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/node2.log):
```
INFO  2023-10-15 09:02:43,732 [shard 0:stre] storage_service - bootstrap[dfa4d19e-229a-4670-b154-d3e0196379fc]: Marked ops done from coordinator=127.0.39.4
INFO  2023-10-15 09:02:44,404 [shard 0:goss] storage_service - Set host_id=74df4b09-7f6e-43e4-9bbe-88ce5962f30d to be owned by node=127.0.39.4
```

node3 (https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/1697360579238_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/node3.log):
```
INFO  2023-10-15 09:02:43,732 [shard 0:stre] storage_service - bootstrap[dfa4d19e-229a-4670-b154-d3e0196379fc]: Marked ops done from coordinator=127.0.39.4
INFO  2023-10-15 09:02:44,517 [shard 0:goss] storage_service - Set host_id=74df4b09-7f6e-43e4-9bbe-88ce5962f30d to be owned by node=127.0.39.4
```
node4 ():
```
INFO  2023-10-15 09:02:43,761 [shard 0:main] init - starting auth service
INFO  2023-10-15 09:02:43,762 [shard 0:main] init - starting batchlog manager
INFO  2023-10-15 09:02:43,762 [shard 0:main] init - starting load meter
INFO  2023-10-15 09:02:43,762 [shard 0:main] init - starting cf cache hit rate calculator
INFO  2023-10-15 09:02:43,762 [shard 0:main] init - starting view update backlog broker
INFO  2023-10-15 09:02:43,762 [shard 0:main] init - Waiting for gossip to settle before accepting client requests...
INFO  2023-10-15 09:02:43,816 [shard 1:stre] compaction - [Reshape ks.cf 9a0f5cb0-6b39-11ee-b9a4-21e2e7d76970] Reshaped 170 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-jx9civlk/test/node4/data/ks/cf-7db1db606b3911eebba4a2c01d5245a0/me-3ga8_0p4j_47yog2tmpv0ba5sc4w-big-Data.db:level=0]. 914kB to 147kB (~16% of original) in 21ms = 43MB/s. ~21760 total partitions merged to 511.
INFO  2023-10-15 09:02:43,816 [shard 1:stre] compaction_manager - Done with off-strategy compaction for ks.cf compaction_group=0/1
INFO  2023-10-15 09:02:43,879 [shard 0:stre] compaction - [Reshape ks.cf 9a129100-6b39-11ee-a5b8-21e1e7d76970] Reshaped 169 sstables to [/jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-jx9civlk/test/node4/data/ks/cf-7db1db606b3911eebba4a2c01d5245a0/me-3ga8_0p4j_4d3v42iq34sjgl83n4-big-Data.db:level=0]. 908kB to 148kB (~16% of original) in 52ms = 17MB/s. ~21632 total partitions merged to 507.
INFO  2023-10-15 09:02:43,879 [shard 0:stre] compaction_manager - Done with off-strategy compaction for ks.cf compaction_group=0/1
INFO  2023-10-15 09:02:55,766 [shard 0:main] gossip - No gossip backlog; proceeding
INFO  2023-10-15 09:02:55,766 [shard 0:main] init - allow replaying hints
INFO  2023-10-15 09:02:55,766 [shard 0:main] init - Launching generate_mv_updates for non system tables
INFO  2023-10-15 09:02:55,766 [shard 0:main] init - starting the view builder
INFO  2023-10-15 09:02:55,769 [shard 0:main] init - starting native transport
INFO  2023-10-15 09:02:55,769 [shard 0:stat] cql_server_controller - Starting listening for CQL clients on 127.0.39.4:9042 (unencrypted, non-shard-aware)
INFO  2023-10-15 09:02:55,769 [shard 0:stat] cql_server_controller - Starting listening for CQL clients on 127.0.39.4:19042 (unencrypted, shard-aware)
INFO  2023-10-15 09:02:55,770 [shard 0:main] init - serving
INFO  2023-10-15 09:02:55,770 [shard 0:main] init - Scylla version 5.4.0~dev-0.20231015.055f0617064d initialization completed.
INFO  2023-10-15 09:02:56,210 [shard 0:main] compaction_manager - Asked to stop
```

So simple timeout issue? doesn't make sense...

**Comment by mykaul** [comment]
> If CL=ONE, why do we wait for two responses? I guess RF=1 and we're bootstrapping a new node.

@bhalevy had a potentially similar issue @ https://github.com/scylladb/scylladb/issues/15454#issuecomment-1763321337

**Comment by avikivity** [comment]
@fruch what's the concurrency of dtest? how many tests, on how many vcpus?

Are we certain that it's running on NVMe? That we use --kernel-page-cache, --unsafe-bypass-fsync, and --overprovisioned?

**Comment by fruch** [comment]
> @fruch what's the concurrency of dtest? how many tests, on how many vcpus?

5 concerencece tests, each node by default is 2 vcpu, and that it, we don't do any housekeeping to limit the tests.

> 
> Are we certain that it's running on NVMe? That we use --kernel-page-cache, --unsafe-bypass-fsync, and --overprovisioned?

NVMe we are sure as for commandline argument I'm quite sure, but I'll check again.

**Comment by fruch** [comment]
> > @fruch what's the concurrency of dtest? how many tests, on how many vcpus?
> 
> 5 concerencece tests, each node by default is 2 vcpu, and that it, we don't do any housekeeping to limit the tests.
> 
> > 
> > Are we certain that it's running on NVMe? That we use --kernel-page-cache, --unsafe-bypass-fsync, and --overprovisioned?
> 
> NVMe we are sure as for commandline argument I'm quite sure, but I'll check again.
> 
> 

This is the commandline

jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-jx9civlk/test/node3/bin/scylla --options-file /jenkins/workspace/scylla-master/gating-dtest-release/scylla/.dtest/dtest-jx9civlk/test/node3/conf/scylla.yaml --log-to-stdout 1 --abort-on-seastar-bad-alloc --abort-on-lsa-bad-alloc 1 --abort-on-internal-error 1 --api-address 127.0.39.3 --collectd-hostname ip-10-0-2-59.ec2.internal.node3 --smp 2 --memory 1024M --developer-mode true --default-log-level info --collectd 0 --overprovisioned --prometheus-address 127.0.39.3 --unsafe-bypass-fsync 1 --kernel-page-cache 1 --commitlog-use-o-dsync 0 --max-networking-io-control-blocks 1000"

So yes, those are used

**Comment by avikivity** [comment]
> 
> NVMe we are sure as for commandline argument I'm quite sure, but I'll check again.


https://github.com/scylladb/scylla-pkg/issues/3620

**Comment by ptrsmrn** [comment]
> If CL=ONE, why do we wait for two responses? I guess RF=1 and we're bootstrapping a new node.

@avikivity yes, as Benny pointed out (in [comment](https://github.com/scylladb/scylladb/issues/15454#issuecomment-1763321337) quoted by @mykaul):

> It looks like the write errors happen while node4 is bootstrapping, while we temporarily increase CL by one to ensure quorum in both old and new replica set:

though I failed to find info on bumping CL when adding a node to a cluster in our doc: https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html
BTW. the doc mentioned that nodes have to be up before adding a new node, which seems to be the case when looking at per-node logs. Again, I'll try to reproduce it locally overnight, this time with trace level logs enabled.

**Comment by bhalevy** [comment]
> though I failed to find info on bumping CL when adding a node to a cluster in our doc: https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html

@ptrsmrn incrementing the required number of nodes is internal to scylla, the application does not need to pass a different CL.

See https://github.com/scylladb/scylladb/blob/055f0617064d20e5735dcff48362508627441b0c/service/storage_proxy.cc#L1330

**Comment by bhalevy** [comment]
> Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6663/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_add_new_node_while_add_new_table_during_bootstrapping/

This looks like a schema propagation issue.

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/1697360579238_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/node2.log
```
INFO  2023-10-15 09:02:34,030 [shard 0:stat] schema_tables - Creating ks1.cf1 id=944a76c0-6b39-11ee-9e80-0e4b7073f6c3 version=634536fa-a121-3ffe-9e3e-0cffa601dc07
INFO  2023-10-15 09:02:34,036 [shard 0:stat] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
```

But, on https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5228/artifact/logs-full.release.006/1697360579238_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_add_new_node_while_add_new_table_during_bootstrapping/node4.log
```
WARN  2023-10-15 09:02:35,272 [shard 0:stat] storage_proxy - Failed to apply mutation from 127.0.39.2#0: std::_Nested_exception<schema_version_loading_failed> (Failed to load schema version 634536fa-a121-3ffe-9e3e-0cffa601dc07): data_dictionary::no_such_keyspace (Can't find a keyspace ks1)

and the table is created locally only later

INFO  2023-10-15 09:02:42,388 [shard 0:stat] schema_tables - Creating keyspace ks1
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.2:0
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.2:0
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.1:0
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.1:0
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.3:0
INFO  2023-10-15 09:02:42,388 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.3:0
INFO  2023-10-15 09:02:42,389 [shard 1:stre] migration_manager - Requesting schema pull from 127.0.39.3:0
INFO  2023-10-15 09:02:42,389 [shard 1:stre] migration_manager - Pulling schema from 127.0.39.3:0
INFO  2023-10-15 09:02:42,394 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.1:0
INFO  2023-10-15 09:02:42,397 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.3:0
INFO  2023-10-15 09:02:42,399 [shard 0:goss] migration_manager - Requesting schema pull from 127.0.39.2:0
INFO  2023-10-15 09:02:42,401 [shard 0:stre] migration_manager - Requesting schema pull from 127.0.39.3:0
INFO  2023-10-15 09:02:42,410 [shard 0:stat] schema_tables - Schema version changed to 733da6b5-42dc-34c6-b25c-484af0a4f7be
INFO  2023-10-15 09:02:42,412 [shard 0:stat] schema_tables - Creating ks1.cf1 id=944a76c0-6b39-11ee-9e80-0e4b7073f6c3 version=634536fa-a121-3ffe-9e3e-0cffa601dc07
INFO  2023-10-15 09:02:42,421 [shard 0:stat] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,444 [shard 0:goss] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,444 [shard 0:goss] migration_manager - Schema merge with 127.0.39.3:0 completed
INFO  2023-10-15 09:02:42,444 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.3:0
INFO  2023-10-15 09:02:42,459 [shard 0:goss] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,459 [shard 0:goss] migration_manager - Schema merge with 127.0.39.1:0 completed
INFO  2023-10-15 09:02:42,459 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.1:0
INFO  2023-10-15 09:02:42,474 [shard 0:goss] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,474 [shard 0:goss] migration_manager - Schema merge with 127.0.39.2:0 completed
INFO  2023-10-15 09:02:42,474 [shard 0:goss] migration_manager - Pulling schema from 127.0.39.2:0
INFO  2023-10-15 09:02:42,489 [shard 0:stre] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,490 [shard 1:stre] migration_manager - Schema merge with 127.0.39.3:0 completed
INFO  2023-10-15 09:02:42,506 [shard 0:goss] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,507 [shard 0:goss] migration_manager - Schema merge with 127.0.39.3:0 completed
INFO  2023-10-15 09:02:42,527 [shard 0:goss] schema_tables - Schema version changed to b7b336b6-eba1-310e-8ac7-bae1c459f509
INFO  2023-10-15 09:02:42,527 [shard 0:goss] migration_manager - Schema merge with 127.0.39.1:0 completed
```

Could this be related to https://github.com/scylladb/scylladb/issues/15686?
@avikivity @tgrabiec @kbr-scylla @gleb-cloudius

**Comment by bhalevy** [comment]
The bottom line from above is that the `CREATE TABLE` query completes before all nodes are aware of the new table.
I guess this could indeed be related to the new node joining group0, maybe in parallel to joining the ring over gossip, so it may get writes before it's full synchronized over the schema.

**Comment by bhalevy** [comment]
According to https://github.com/scylladb/scylladb/issues/15686#issuecomment-1755608750:
> The topology coordinator generates bootstrap tokens and assigns them to the node, does the transition to bootstrap, only then tells the node to perform the read barrier and read the current system topology/schema state.
> 
> We need to postpone generating and inserting the tokens only after the node does the read barrier.

**Comment by kbr-scylla** [comment]
Dtests are not running in consistent-raft-topology mode.

**Comment by gleb-cloudius** [comment]
> Dtests are not running in consistent-raft-topology mode.

This one runs with raft disabled all together.

**Comment by tgrabiec** [comment]
Looks like https://github.com/scylladb/scylladb/issues/3760

**Comment by kbr-scylla** [comment]
https://github.com/scylladb/scylladb/issues/3760#issuecomment-1638179198
-- I don't think we're planning to fix these issues outside Raft mode.
cc @kostja

**Comment by mykaul** [comment]
If we eventually decide not to fix without raft (not getting if we should or should not), we should ensure the test only runs with raft enabled.

**Comment by kbr-scylla** [comment]
The purpose of using Raft for schema changes was precisely to fix multiple issues with schema changes, and they were. Most of them are not fixable without having some kind of consensus mechanism (like Raft).

Therefore we need to disable such tests outside Raft mode.

**Comment by bhalevy** [comment]
> If we eventually decide not to fix without raft (not getting if we should or should not), we should ensure the test only runs with raft enabled.

> Therefore we need to disable such tests outside Raft mode.

But if we don't fix it and not test it, we're bound to hitting it in the field.

Maybe our support matrix is just too big.
We have:
- schema changes with or without raft, cross
- topology change with or without raft, cross
- repair based vs. streaming based node operations (unrelated to this issue)
- and soon, vnodes vs. tablets

Maybe it's time to prune the tree and stop supporting (and testing) some combinations.
So raft all the way (schema and topology changes) vs. raft disabled both for schema changes and topology changes.

> This one runs with raft disabled all together.

And this combination should still be supported and have no regressions, until we deprecate and remove support for disabling raft.

**Comment by gleb-cloudius** [comment]
> 
> > This one runs with raft disabled all together.
> 
> And this combination should still be supported and have no regressions, until we deprecate and remove support for disabling raft.

And this is not a regression, no?

**Comment by bhalevy** [comment]
> And this is not a regression, no?

The frequency we hit it recently is definitely a regression.
From a theoretical problem it became a practical one.

**Comment by tgrabiec** [comment]
What exposes the problem seems to be slow propagation to schema change RPC to the bootstrapping node. The bootstrapping node gets the schema 10 seconds later. If propagation is fast, the standard schema agreement wait should cover the bootstrapping node as well.

**Comment by avikivity** [comment]
It could be the infrastructure issue (running on EBS GP2). It slows down everything, and then gossip gives up and moves on.

**Comment by mykaul** [comment]
Assuming slow infra exposed this problem. While indeed an issue, it is clearly not going to be fixed in 5.4. Pushing to backlog.
